Plugin Framework: propagate logging output to Pulumi

t0yv0 opened the pull request:
Fixes #820 

Dependencies (required upstream fixes to pass tests):

github.com/pulumi/pulumi/pkg/v3 v3.58.0
github.com/pulumi/pulumi/sdk/v3 v3.58.0


Currently this gives output that looks something like this through Pulumi CLI:

```
  random:index:RandomInteger (integer):
    2023-03-15T15:08:51.429-0500  provider/resource_integer.go:100: provider: Create RandomInteger - INFO: provider=random providerVersion="4.12.0-alpha.1678477401+fa71af56.dirty" superfield=supervalue
    warning: 2023-03-15T15:08:51.430-0500  provider/resource_integer.go:101: provider: Create RandomInteger - WARN: provider=random providerVersion="4.12.0-alpha.1678477401+fa71af56.dirty" superfield=supervalue
    error: 2023-03-15T15:08:51.430-0500 provider/resource_integer.go:102: provider: Create RandomInteger - ERROR: providerVersion="4.12.0-alpha.1678477401+fa71af56.dirty" superfield=supervalue provider=random
    2023-03-15T15:08:51.430-0500  provider/resource_integer.go:111: provider: Create RandomInteger - INFO +fields: b=b a=1 provider=random providerVersion="4.12.0-alpha.1678477401+fa71af56.dirty" superfield=supervalue
    warning: 2023-03-15T15:08:51.431-0500  provider/resource_integer.go:112: provider: Create RandomInteger - WARN +fields: superfield=supervalue a=1 b=b provider=random providerVersion="4.12.0-alpha.1678477401+fa71af56.dirty"
    error: 2023-03-15T15:08:51.431-0500 provider/resource_integer.go:113: provider: Create RandomInteger - ERROR +fields: superfield=supervalue provider=random providerVersion="4.12.0-alpha.1678477401+fa71af56.dirty" a=1 b=b
```

- What should be the default log level? Should it be INFO as in PR or should it be OFF by default until the user sets TF_LOG?
- Should we include timestamps or drop?
- Do we like source locators like provider/resource_integer.go:113 ? Seems somewhat useful for debugging 
- How about provider=random and providerVersion="..." - include or drop?
- Any user-provided custom fields like a=1 should probably stay

t0yv0 commented on pf/internal/logging/logging_test.go:
```diff
@@ -0,0 +1,161 @@
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+func TestLogging(t *testing.T) {
```
This is the interesting new bit, new package to connect up logging and the tests thereof.

iwahbe commented on pf/internal/logging/logging.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+func parseTfLogEnvVar() hclog.Level {
+	env := os.Getenv("TF_LOG")
+	if env == "" {
```
I think this reads better, but your free to discard.
```suggestion
	env, present := os.LookupEnv("TF_LOG")
	if !present {
```

You could also just remove the `env == ""` check, letting `default:` take care of it.

iwahbe reviewed (approved):
> What should be the default log level? Should it be INFO as in PR or should it be OFF by default until the user sets TF_LOG?

Since we will be consuming logs generated by providers written for TF, I think we should emulate what TF does. 

> Should we include timestamps or drop?

I don't generally find them useful, but no strong feelings either way.

> Do we like source locators like provider/resource_integer.go:113 ? Seems somewhat useful for debugging

Love it

> How about provider=random and providerVersion="..." - include or drop?

I think we should include, but lets shorten them to the way the plugin loader logs this:

Replace

```
provider=random providerVersion="4.12.0 alpha.1678477401+fa71af56.dirty"
```

With

```
provider=random@4.12.0alpha.1678477401+fa71af56.dirty
```

>Any user-provided custom fields like a=1 should probably stay

🆗

t0yv0 commented:
So it turns out if one follows instructions from Pulumi troubleshooting page, the logs are written out to a file already, and Pulumi adds timestamps, so that's great. I removed timestamps at TF level. Addressed the other bits. Will merge once I get the tests to pass.
